mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -394,181 +393,134 @@ struct BlockwiseGemmV2RewritePattern
-      auto ilnkiv = innerLoopNK.getInductionVar();
-
-      Value sourceOffsetB = ilnkb.create<AddIOp>(
-          loc,
-          ilnkb.create<AddIOp>(
-              loc, ilnkb.create<MulIOp>(loc, ilnkiv, NConstantOp), laneId),
-          nOffset);
-
-      if (KPack > 1)
-        sourceOffsetB = ilnkb.create<MulIOp>(
-            loc, sourceOffsetB, ilnkb.create<ConstantIndexOp>(loc, KPack));
-
-      auto destOffsetB = ilnkb.create<AddIOp>(loc, ilnkiv, kOffsetB);
-
-      Value valueB = ilnkb.create<InBoundsLoadOp>(
-          loc, bufferBElementType, op.getMatrixB(), sourceOffsetB);
-      ilnkb.create<memref::StoreOp>(loc, valueB, bufferB,
-                                    ValueRange{destOffsetB});
-    } else {
-      // const index_t blk_id = laneId / mfma_type.num_threads_blk;
-      // const index_t blk_td = laneId % mfma_type.num_threads_blk;
-      auto inputSpanLenConstantOp =
-          b.create<ConstantIndexOp>(loc, inputSpanLen);
-      auto blk_id = b.create<DivUIOp>(loc, laneId, inputSpanLenConstantOp);
-      auto blk_td = b.create<RemUIOp>(loc, laneId, inputSpanLenConstantOp);
-
-      Value kBaseA = b.create<AddIOp>(loc, aBase, blk_td);
-      Value kBaseB = b.create<AddIOp>(loc, bBase, blk_td);
-
-      // for(index_t k_i = 0; k_i < KPerThread; k_i += mfma_type.num_input_blks)
-      // {
-      //     a[k_i] = p_a_wave[(k_i * num_input_blks + blk_id) * M + blk_td];
-      //     b[k_i] = p_b_wave[(k_i * num_input_blks + blk_id) * N + blk_td];
-      // }
-      // p_a_wave need to be offseted by waveOffsetA.
-      // p_b_wave need to be offseted by waveOffsetB.
-
-      auto inputSpansPerMfmaInConstantOp =
-          b.create<ConstantIndexOp>(loc, inputSpansPerMfmaIn);
-
-      auto loopKLoad = b.create<AffineForOp>(loc, 0, KPerThread);
-      auto lklb = ConversionPatternRewriter::atBlockBegin(loopKLoad.getBody(),
-                                                          b.getListener());
-      auto lkliv = loopKLoad.getInductionVar();
-
-      Value sourceOffsetA = lklb.create<AddIOp>(
-          loc,
-          lklb.create<MulIOp>(
-              loc,
-              lklb.create<AddIOp>(
-                  loc,
-                  lklb.create<MulIOp>(loc, lkliv,
-                                      inputSpansPerMfmaInConstantOp),
-                  blk_id),
-              MConstantOp),
-          kBaseA);
-
+    Value KPerThreadConstantOp = b.create<ConstantIndexOp>(loc, KPerThread);
+
+    // store bufferA logic.
+    // for(index_t m_i = 0; n_i < mRepeats; ++n_i)
+    //   for(index_t k_i= 0; k_i < K; ++k_i)
+    //
+    // Note: p_a_wave need to be offseted by waveOffsetA.
+    Value inputSpanLenConstantOp = b.create<ConstantIndexOp>(loc, inputSpanLen);
+    Value inputSpansPerMfmaInConstantOp =
+        b.create<ConstantIndexOp>(loc, inputSpansPerMfmaIn);
+
+    Value blk_id = b.create<DivUIOp>(loc, laneId, inputSpanLenConstantOp);
+    Value blk_td = b.create<RemUIOp>(loc, laneId, inputSpanLenConstantOp);
+
```

Comment:
Move those in the reduction condition to scope them better.

Reply:
I did this, but a thought occurred to me. If we bring the `blk_id` and `blk_td` inside the `lambda` we are duplicating these. I.e., we will have:
```
blkid = computeblkid()
blktd = computeblktd()
for m:
    for k:

blkid = computeblkid()
blktd = computeblktd()
for n:
    for k:
```
Is this ok? I am worried that the compiler might not be so smart to get rid of the duplication

Reply:
cc also @krzysz00

Reply:
I _think_ those operations have `NoSideEffects`/`Pure`, so when we run CSE, which we eventually will, they'll get merged

Reply:
Yes they do (just checked) :) Thanks @krzysz00 !